openpower/patches/vesnin-patches/petitboot/petitboot-1001-discover-platform-powerpc-fix-bug-with-ipmi-boot-mai.patch

```diff
@@ -0,0 +1,280 @@
++++ b/discover/platform-powerpc.c
+@@ -22,6 +22,29 @@
+ #include "ipmi.h"
+ #include "dt.h"
+ 
++#define CHASSIS_BOOT_MBOX_IANA_SZ 3
++#define CHASSIS_BOOT_MBOX_DATA_SZ 16
++#define CHASSIS_BOOT_MBOX_BLOCK0_DATA_SZ \
++	(CHASSIS_BOOT_MBOX_DATA_SZ - CHASSIS_BOOT_MBOX_IANA_SZ)
++
++typedef struct __attribute__((packed)) {
++	uint8_t iana[CHASSIS_BOOT_MBOX_IANA_SZ];
++	uint8_t data[CHASSIS_BOOT_MBOX_BLOCK0_DATA_SZ];
++} mbox_block0_t;
++
++typedef union {
++		uint8_t data[CHASSIS_BOOT_MBOX_DATA_SZ];
++		mbox_block0_t b0;
++} mbox_t;
++
++typedef struct __attribute__((packed)) {
++	uint8_t cc;
++	uint8_t param_version;
++	uint8_t param_valid;
++	uint8_t block_selector;
++	mbox_t mbox;
++} ipmi_mbox_response_t;
++
+ static const char *partition = "common";
+ static const char *sysparams_dir = "/sys/firmware/opal/sysparams/";
+ static const char *devtree_dir = "/proc/device-tree/";
+@@ -437,10 +460,10 @@ static int get_ipmi_bootdev_ipmi(struct platform_powerpc *platform,
+ }
+ 
+ static int get_ipmi_boot_mailbox_block(struct platform_powerpc *platform,
+-		char *buf, uint8_t block)
++		mbox_t *mailbox, uint8_t block)
+ {
+-	size_t blocksize = 16;
+-	uint8_t resp[3 + 16];
++	size_t blocksize;
++	ipmi_mbox_response_t ipmi_mbox_resp;
+ 	uint16_t resp_len;
+ 	char *debug_buf;
+ 	int rc;
+@@ -450,59 +473,68 @@ static int get_ipmi_boot_mailbox_block(struct platform_powerpc *platform,
+ 		0x00,  /* no block selector */
+ 	};
+ 
+-	resp_len = sizeof(resp);
++	resp_len = sizeof(ipmi_mbox_response_t);
+ 	rc = ipmi_transaction(platform->ipmi, IPMI_NETFN_CHASSIS,
+ 			IPMI_CMD_CHASSIS_GET_SYSTEM_BOOT_OPTIONS,
+ 			req, sizeof(req),
+-			resp, &resp_len,
++			(uint8_t*)&ipmi_mbox_resp, &resp_len,
+ 			ipmi_timeout);
+ 	if (rc) {
+ 		pb_log("platform: error reading IPMI boot options\n");
+ 		return -1;
+ 	}
+ 
+-	if (resp_len < sizeof(resp)) {
+-		if (resp_len < 3) {
+-			pb_log("platform: unexpected length (%d) in "
+-					"boot options mailbox response\n",
+-					resp_len);
+-			return -1;
+-		}
+-
+-		if (resp_len == 4) {
+-			pb_debug_fn("block %hu empty\n", block);
+-			return 0;
+-		}
++	if (resp_len > sizeof(ipmi_mbox_response_t)) {
++		pb_debug("platform: invalid mailbox response size!\n");
++		return -1;
++	}
+ 
+-		blocksize = sizeof(resp) - 3;
+-		pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
+-				block, blocksize);
++	if (resp_len < 3) {
++		pb_log("platform: unexpected length (%d) in "
++				"boot options mailbox response\n",
++				resp_len);
++		return -1;
+ 	}
+ 
+-	debug_buf = format_buffer(platform, resp, resp_len);
++	blocksize = resp_len - (sizeof(ipmi_mbox_response_t) - sizeof(mbox_t));
++	pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
++			block, blocksize);
++
++	debug_buf = format_buffer(platform, (uint8_t*)&ipmi_mbox_resp, resp_len);
+ 	pb_debug_fn("IPMI bootdev mailbox block %hu:\n%s\n", block, debug_buf);
+ 	talloc_free(debug_buf);
+ 
+-	if (resp[0] != 0) {
++	if (ipmi_mbox_resp.cc != 0) {
```

Comment:
I assume, they don't have macros like `IPMI_CC_OK` in petitboot's shitcode?

Reply:
Yes you are right. All macros for ipmi functions are in the https://github.com/open-power/petitboot/blob/master/discover/ipmi.h